### DISCUSSION TITLE
Why is Distributed locking needed?

### DISCUSSION THREAD

**Discussion by tomy2105** [elsa-workflows/elsa-core#2605] [Q&A] [answered]
My question is probably a result of my lack of understanding about Elsa internals (how are workflow bursts distributed between multiple Elsa instances), so apologies in advance about that.

Can someone explain to me, in a bit of detail, why is distributed locking on workflow instance needed for every workflow burst.
What are we preventing with this lock? 

I ask because I made following assumptions (and one of them is probably wrong) about workflow start/resume:
- if started/resumed as a result of HTTP request, request will be served to a single instance so don't see what other instances would race for it
- if started/resumed by timer/cron/etc.... I assume Hangfire/Quartz if configured in clustering mode should be able to trigger it on single instance only
- if started/resumed by same message queueing mechanism it should be a mq configuration (one to one, one to many) to ensure delivery to single or multiple instances as required.
- etc.....

So if you could in short term explain Elsa internals regarding how pending bursts of workflow are distributed between multiple instances and what scenario is distributed lock on instance id protecting against?

Just to explain why I ask this. When Distributed lock library is configured with Postgres backend it seems to create one additional per burst executed db connection for the lock. And it is kept open for the whole length of workflow burst, and it does nothing else on this connection which is a big "waste" of resources unless it is absolutely needed. In addition total connection count a database can handle imposes a limit on number of concurrent bursts that can be executed, so there is a scalability problem too.

Thank you in advance.

**Comment by sfmskywalker** [comment, marked as answer]
The distributed lock is needed in when you are hosting your workflow server in a multi-node cluster (e.g. Kubernetes or Azure App Service with multiple nodes).

Here's why:

Imagine a workflow that is blocked on a **Timer** activity as well as a **Message Received** activity. If the Timer activity is triggered, the system will load the workflow instance into memory on e.g. Node 1 in the cluster and start executing each subsequent activity connected to the timer until the next blocking activity or the end of the workflow is reached (aka "burst of execution").

Imagine next that during this burst of execution, a message is received from some service bus, and gets picked up on Node 2 in the cluster. Node 2 will also load the workflow instance in memory (possible an outdated version, if Node 1 has not yet stored the changed workflow instance in the DB) and begin a burst of execution.

When these two bursts of executions occur in parallel, a few things can go wrong:

1. Concurrency issues: one workflow instance in memory is stale compared to the other one. The last one that gets persisted "wins", overwriting the other one. In the above example, this may mean that the workflow will be blocked on the Message Received activity again.
2. Incomplete state: if you have workflows with parallel execution paths, they may be updating variables that other paths depend on. But as with the first issue, this state can get lost if the last executed workflow instance overwrites the previous one.
3. Race conditions: imagine having a workflow with a trigger and a blocking activity somewhere down the line, like a **Signal Received**. If a signal is sent on Node 2 **while the workflow instance is still in memory on Node 1**, the signal will be lost, since the workflow instance isn't blocked yet on the Signal Received activity. However, with distributed locking in place, **Node 2 will wait while it tries to acquire a lock on the same workflow instance** - allowing Node 2 to deliver the signal after the workflow instance lock held by Node 1 is released. This is also why it is important that bursts of execution happen as fast as possible to decrease the amount of time a distributed lock is held.

To solve potentially strange & seemingly unexpected effects, it is important that only one thread on a single machine executes a workflow instance at any one time.

To achieve that, there are two useful strategies (that I'm aware of):

1. Distributed locking - works, but diminished throughput due to nodes having to wait when a lock is already taken.
2. Actor model / inbox per workflow instance - ideal for massive throughput - no distributed locking necessary.

When a burst of execution takes place, it always happens on the same machine in the same thread. Once the burst ends (because a blocking activity was encountered or the end of the workflow), the workflow instance gets persisted and goes out of memory.

> if started/resumed as a result of HTTP request, request will be served to a single instance so don't see what other instances would race for it

Correct, only one node will receive this request, and that node will execute the workflow instance until a blocking activity or end of workflow is encountered.

If all of your workflows are simple ones where there are no blocking activities, then distributed locking isn't necessary and you can disable it. But as soon as you have a workflow with a blocking activity, you should use distributed locking (or the actor model, which is available with Elsa 3 as the default runtime model).

> * if started/resumed by timer/cron/etc.... I assume Hangfire/Quartz if configured in clustering mode should be able to trigger it on single instance only

Exactly.

> * if started/resumed by same message queueing mechanism it should be a mq configuration (one to one, one to many) to ensure delivery to single or multiple instances as required.

Yep.


> And it is kept open for the whole length of workflow burst, and it does nothing else on this connection which is a big "waste" of resources unless it is absolutely needed

Indeed. To minimize that, try to keep your activities fast. If you find that some activities take more time to execute because of some heavy lifting, consider doing that work from a background job (e.g. using Hangfire), while the workflow instance goes to sleep. Once the job is done, it can resume the workflow. I had to do this for a customer who has a set of custom activities making HTTP calls to Azure Functions. Some of these functions take several minutes to complete.

> In addition total connection count a database can handle imposes a limit on number of concurrent bursts that can be executed, so there is a scalability problem too

Exactly, the distributed locking mechanism ensures data consistency but limits throughput and therefore the ability to scale. A workaround might be to add a Redis server to your infrastructure to reduce the number of connections to your Postgres DB.
There would still be a limit on your apps scalability of course, but it should at least increase.

The best solution is to change the architecture such that every workflow instance gets its own "worker" that feeds off a queue (aka known as "inbox" in this context). As I mentioned, Elsa 3 will have this as a default model.

**Reply by tomy2105** [reply]
Again, thank you very much for great library and great and explanatory answer @sfmskywalker!

Let me see if I understood the core of your answer correctly.....

Distributed locking prevents havoc happening which might happen if you have more than **_two_ blocking activities** in a workflow **active at the same time**. Because then they could be resumed at a same _or similar_ time which might cause race condition, stale data, etc... all the bad stuff. 

I forgot about this possibility in Elsa because I think I'm not using it, but let me check just to be sure...
If I only have **one** (or zero :)) blocking activity active at the same time I don't need distributed locking?
**Am I correct so far?**

If I don't have fork activity enabled, I think I cannot have two blocking activities active at the same time, so I don't need distributed locking?
_Is this correct?
Is there anything else besides fork activity that can produce parallel execution so multiple blocking activities could be active at the same time?_

**Reply by sfmskywalker** [reply]
Yep, you got. 

But it doesn't necessarily require a fork activity to go wrong - it could even be problematic if you have just a single blocking activity (not to be confused with a trigger activity, which is a starting activity) as described in my 3rd example with race conditions where the workflow has started but not yet reached the blocking activity, and yet some event happens that would normally resume that blocking activity.

If you only have workflows in your system that finish from start to end then there's no need for distributed locking.

**Reply by tomy2105** [reply]
I at the moment (who knows that will happen in future :)) I have http trigger (starting) activities and Timer (with delay in seconds, no cron, repeat, etc...) as blocking one inside workflow.
Since timer without repeat should never trigger twice am I safe to assume I don't need distributed locking?
Provided my timer timeout is large enough that I'm sure current node will save it state to DB before timer triggers again.

**Reply by sfmskywalker** [reply]
Probably safe. If I'm not mistaken, Quartz.NET or Hangfire will make sure that the timer job executes only on one node in the cluster, which means only one node will be executing that particular workflow instance being resumed.

**Reply by tomy2105** [reply]
Thanks again for in depth explanation :).

**Reply by mickdelaney** [reply]
a kafka bus could potentially help here if the Partition key was the WF correlation id. that would ensure that the message is handled in an ordered fashion by a single consumer group. sort of like an Actor.

**Reply by sfmskywalker** [reply]
Indeed. Kafka might make for a good additional implementation of the runtime engine for Elsa 3 (which currently uses [ProtoActor](https://proto.actor/)).
